File: packages/ts/core/src/middleware/resilience.ts
import { ProviderError, RateLimitError, TimeoutError } from "../errors.js";

/**
 * Options for {@link withResilience}. `maxRetries` + `timeoutMs` are required
 * (the orchestrator sources them from `ProviderConfig`); the rest have defaults.
 * `random` is injectable so tests can make jitter deterministic.
 */
export interface ResilienceOptions {
  maxRetries: number;
  timeoutMs: number;
  baseDelayMs?: number;
  maxDelayMs?: number;
  idempotencyKey?: string;
  /** External abort that, when fired, also aborts the in-flight attempt. */
  signal?: AbortSignal;
  random?: () => number;
}

const DEFAULT_BASE_DELAY_MS = 100;
const DEFAULT_MAX_DELAY_MS = 10_000;

/** Context handed to the wrapped thunk on every attempt (stable idempotency key). */
export interface ResilienceCallContext {
  idempotencyKey?: string;
  signal: AbortSignal;
}

const sleep = (ms: number): Promise<void> =>
  new Promise<void>((resolve) => setTimeout(resolve, ms));

/**
 * Retriable iff the error is part of the Feature 02 taxonomy that the design
 * marks safe to replay: a rate limit, a timeout, or any `ProviderError` whose
 * `retriable` flag is set (e.g. a mapped 5xx). Everything else — auth failures,
 * non-retriable provider errors — propagates immediately.
 */
export function isRetriable(err: unknown): boolean {
  if (err instanceof RateLimitError || err instanceof TimeoutError) return true;
  if (err instanceof ProviderError) return err.retriable === true;
  return false;
}

/**
 * Full-jitter exponential backoff: `min(base * 2^attempt, max)` scaled by a
 * random factor in [0, 1). `attempt` is 0-based, so the first retry waits up to
 * `base`, the next up to `2*base`, capped at `maxDelayMs`.
 */
export function computeBackoffMs(
  attempt: number,
  opts: ResilienceOptions,
): number {
  const base = opts.baseDelayMs ?? DEFAULT_BASE_DELAY_MS;
  const max = opts.maxDelayMs ?? DEFAULT_MAX_DELAY_MS;
  const random = opts.random ?? Math.random;
  return Math.min(base * 2 ** attempt, max) * random();
}

/**
 * Wrap the async thunk the orchestrator uses to invoke the resolved provider,
 * making it resilient *once* in `core` rather than in every adapter:
 *
 * - Each attempt is bounded by `timeoutMs` via an `AbortController`; on expiry
 *   the attempt is aborted and a `core` {@link TimeoutError} is raised (then
 *   itself subject to retry).
 * - Retriable errors are retried with exponential backoff + full jitter, capped
 *   by `maxRetries`; non-retriable errors propagate on the first attempt.
 * - The `idempotencyKey` is threaded unchanged into every attempt's context so
 *   retried requests are safe to replay.
 *
 * Provider-agnostic: it imports only the error taxonomy and native timers/abort
 * primitives — no adapter or SDK.
 */
export async function withResilience<T>(
  call: (ctx: ResilienceCallContext) => Promise<T>,
  opts: ResilienceOptions,
): Promise<T> {
  let lastErr: unknown;
  for (let attempt = 0; attempt <= opts.maxRetries; attempt++) {
    const ac = new AbortController();
    const onExternalAbort = () => ac.abort();
    // An already-fired or later-fired external signal aborts this attempt too.
    if (opts.signal) {
      if (opts.signal.aborted) ac.abort();
      else opts.signal.addEventListener("abort", onExternalAbort, { once: true });
    }
    const timer = setTimeout(() => ac.abort(), opts.timeoutMs);
    try {
      return await call({ idempotencyKey: opts.idempotencyKey, signal: ac.signal });
    } catch (err) {
      // Distinguish a deadline abort from a provider-thrown error: a fired
      // abort means we hit the timeout, so surface a core TimeoutError.
      lastErr =
        ac.signal.aborted && !(err instanceof TimeoutError)
          ? new TimeoutError(`request exceeded ${opts.timeoutMs}ms`, { cause: err })
          : err;
      if (attempt < opts.maxRetries && isRetriable(lastErr)) {
        await sleep(computeBackoffMs(attempt, opts));
        continue;
      }
      throw lastErr;
    } finally {
      clearTimeout(timer);
      opts.signal?.removeEventListener("abort", onExternalAbort);
    }
  }
  // Unreachable: the loop either returns or throws, but satisfies the type checker.
  throw lastErr;
}

File: packages/py/llmorch-core/src/llmorch_core/middleware/resilience.py
"""Resilience middleware — twin of packages/ts/core/src/middleware/resilience.ts.

`with_resilience` wraps the coroutine the orchestrator uses to invoke the
resolved provider, making it resilient *once* in core rather than in every
adapter: exponential backoff + full jitter on retriable errors only, an
``asyncio.wait_for`` per-attempt deadline that surfaces a core ``TimeoutError``,
and an idempotency key threaded unchanged across attempts. It imports only the
error taxonomy and the stdlib ``asyncio``/``random`` — no adapter or SDK.
"""

import asyncio
import random as _random
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from typing import Any, TypeVar

from ..errors import ProviderError, RateLimitError, TimeoutError

T = TypeVar("T")

# The context dict handed to the wrapped call on every attempt.
CallContext = dict[str, Any]


@dataclass
class ResilienceOptions:
    """Tuning for :func:`with_resilience`.

    ``max_retries`` + ``timeout_ms`` are sourced by the orchestrator from
    ``ProviderConfig``; the rest have defaults. ``rng`` is injectable so tests
    can make jitter deterministic.
    """

    max_retries: int
    timeout_ms: int
    base_delay_ms: int = 100
    max_delay_ms: int = 10_000
    idempotency_key: str | None = None
    rng: Callable[[], float] = field(default=_random.random)


def _is_retriable(err: BaseException) -> bool:
    """Retriable iff a rate limit, a timeout, or a ``ProviderError`` so flagged."""
    if isinstance(err, (RateLimitError, TimeoutError)):
        return True
    return isinstance(err, ProviderError) and bool(getattr(err, "retriable", False))


def _compute_backoff(attempt: int, opts: ResilienceOptions) -> float:
    """Full-jitter exponential backoff in *seconds* (0-based ``attempt``)."""
    # `int ** int` is typed `Any` in typeshed (negative exponents yield float),
    # so pin the arithmetic to float explicitly.
    ceiling: float = min(opts.base_delay_ms * 2**attempt, opts.max_delay_ms)
    return ceiling * opts.rng() / 1000.0


async def with_resilience(
    call: Callable[[CallContext], Awaitable[T]],
    opts: ResilienceOptions,
) -> T:
    """Drive ``call`` to success through backoff retries, or raise the final error.

    Each attempt is bounded by ``timeout_ms`` via ``asyncio.wait_for``; an expiry
    maps the ``asyncio.TimeoutError`` to a core ``TimeoutError`` (then itself
    subject to retry). Retriable errors are retried with jittered exponential
    backoff capped by ``max_retries``; everything else propagates immediately.
    """
    last_err: BaseException | None = None
    for attempt in range(opts.max_retries + 1):
        ctx: CallContext = {"idempotency_key": opts.idempotency_key}
        try:
            return await asyncio.wait_for(call(ctx), timeout=opts.timeout_ms / 1000.0)
        # `asyncio.TimeoutError` is the builtin TimeoutError, which our import of
        # the core TimeoutError shadows — keep the qualified name so this catches
        # the deadline (not our own core error) and remaps it. (ruff UP041 N/A.)
        except asyncio.TimeoutError as e:  # noqa: UP041
            err: BaseException = TimeoutError(f"request exceeded {opts.timeout_ms}ms")
            err.__cause__ = e
        except BaseException as e:  # noqa: BLE001 — re-raised below; we only branch on retriability
            err = e
        last_err = err
        if attempt < opts.max_retries and _is_retriable(err):
            await asyncio.sleep(_compute_backoff(attempt, opts))
            continue
        raise err
    # Unreachable: the loop always returns or raises.
    assert last_err is not None
    raise last_err
